### DISCUSSION TITLE
Help parsing a delimited list

### DISCUSSION THREAD

**Discussion by mzagrabe** [winnow-rs/winnow#401] [Q&A] [answered]
Greetings,

I'm struggling to know the best way to match any characters except a particular one - and also matching to the end of the line.

```
use winnow::PResult;
use winnow::Parser;
use winnow::ascii::{
    space1,
    digit1,
};
use winnow::combinator::separated;
use winnow::token::take_until0;

fn list<'s>(input: &mut &'s str) -> PResult<Vec<&'s str>> {
    separated(
        0..,
        take_until0(";"),
        ";",
    )   
        .parse_next(input)
}

fn main() {
    let mut input = "Game 99: 1 green, 1 red, 12 blue; 2 green, 4 red, 14 blue; 4 blue, 6 red; 10 red, 2 green, 1 blue";
    let parsed = ( 
        "Game",
        space1,
        digit1,
        ":",
        list,
    )   
        .parse_next(&mut input)
        .unwrap()
    ;   

    println!("{parsed:#?}");

    assert_eq!(
        parsed.4,
        vec![
            " 1 green, 1 red, 12 blue",
            " 2 green, 4 red, 14 blue",
            " 4 blue, 6 red",
            " 10 red, 2 green, 1 blue",
        ],
    );  
}
```

`take_until0(";"),` does not feel like the right combinator to use. Also, I'm not successfully matching to the end of the line - so I don't know what the best way to attack this problem. I've tried using `none_of(';')`, but that also was the wrong route.

Any thoughts about my problem?

Thanks for the help!

-m

**Comment by epage** [comment, marked as answer]
You likely want `take_while(|c| c != ';')`.

**Reply by epage** [reply]
Alternative ideas
- wrap `take_until0` in an `alt` with a `rest`
- sequence a `rest` after `separated`

**Comment by mzagrabe** [comment]
Thanks for the help, Ed! Very appreciated.
